staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -222,47 +230,51 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	w.WriteHeader(http.StatusOK)
 	flusher.Flush()
 
-	watchEncoder := newWatchEncoder(req.Context(), kind, s.EmbeddedEncoder, s.Encoder, framer)
-	ch := s.Watching.ResultChan()
-	done := req.Context().Done()
-
-	for {
-		select {
-		case <-s.ServerShuttingDownCh:
-			// the server has signaled that it is shutting down (not accepting
-			// any new request), all active watch request(s) should return
-			// immediately here. The WithWatchTerminationDuringShutdown server
-			// filter will ensure that the response to the client is rate
-			// limited in order to avoid any thundering herd issue when the
-			// client(s) try to reestablish the WATCH on the other
-			// available apiserver instance(s).
-			return
-		case <-done:
-			return
-		case <-timeoutCh:
-			return
-		case event, ok := <-ch:
-			if !ok {
-				// End of results.
+	serveFunc := func() {
```

Comment:
I decided to clean this up a little bit in https://github.com/kubernetes/kubernetes/pull/122115
Let's wait for that PR and adjust on top of that one - I hope it should help a bit with this refactoring.

Reply:
@wojtek-t I rebased on top of https://github.com/kubernetes/kubernetes/pull/122115, which made this refactoring much simpler so we might not need #122060. PTAL.